storage/encode.go

```diff
@@ -1,24 +1,383 @@
 package storage
 
 import (
+	"bytes"
 	"encoding/binary"
 	"fmt"
+	"math"
+	"unsafe"
 
 	"github.com/gnolang/gno/tm2/pkg/amino"
 	"github.com/gnolang/gno/tm2/pkg/bft/types"
 )
 
-// encodeInt64 encodes an int64 value into little endian
-func encodeInt64(value int64) []byte {
-	b := make([]byte, 8)
-	binary.LittleEndian.PutUint64(b, uint64(value))
+const (
+	bytesMarker     byte = 0x12
+	bytesDescMarker      = bytesMarker + 1
 
-	return b
+	// IntMin is chosen such that the range of int tags does not overlap the
+	// ascii character set that is frequently used in testing.
+	IntMin      = 0x80 // 128
+	intMaxWidth = 8
+	intZero     = IntMin + intMaxWidth           // 136
+	intSmall    = IntMax - intZero - intMaxWidth // 109
+	// IntMax is the maximum int tag value.
+	IntMax = 0xfd // 253
+
+	// <term>     -> \x00\x01
+	// \x00       -> \x00\xff
+	escape                   byte = 0x00
+	escapedTerm              byte = 0x01
+	escapedJSONObjectKeyTerm byte = 0x02
+	escapedJSONArray         byte = 0x03
+	escaped00                byte = 0xff
+	escapedFF                byte = 0x00
+)
+
+type escapes struct {
+	escape      byte
+	escapedTerm byte
+	escaped00   byte
+	escapedFF   byte
+	marker      byte
+}
+
+var ascendingBytesEscapes = escapes{escape, escapedTerm, escaped00, escapedFF, bytesMarker}
+
+// EncodeUint32Ascending encodes the uint32 value using a big-endian 4 byte
+// representation. The bytes are appended to the supplied buffer and
+// the final buffer is returned.
+func EncodeUint32Ascending(b []byte, v uint32) []byte {
+	return append(b, byte(v>>24), byte(v>>16), byte(v>>8), byte(v))
 }
 
-// decodeInt64 decodes an int64 value from little endian
-func decodeInt64(value []byte) int64 {
-	return int64(binary.LittleEndian.Uint64(value))
+// EncodeUint32Descending encodes the uint32 value so that it sorts in
+// reverse order, from largest to smallest.
+func EncodeUint32Descending(b []byte, v uint32) []byte {
+	return EncodeUint32Ascending(b, ^v)
+}
+
+// DecodeUint32Ascending decodes a uint32 from the input buffer, treating
+// the input as a big-endian 4 byte uint32 representation. The remainder
+// of the input buffer and the decoded uint32 are returned.
+func DecodeUint32Ascending(b []byte) ([]byte, uint32, error) {
+	if len(b) < 4 {
+		return nil, 0, fmt.Errorf("insufficient bytes to decode uint32 int value")
+	}
+
+	v := binary.BigEndian.Uint32(b)
```

Comment:
I'd say stick to binary.BigEndian methods OR copying the methods (ie. comparing with `EncodeUint32Ascending`, which copies the method instead)